Web Inspector: Elements: show a badge for `<slot>` to quickly jump to the assigned node

dcrousso opened the pull request:
#### 4f88ca91386d7fc52b79472a0f22fd6823ba41fe
<pre>
Web Inspector: Elements: show a badge for `&lt;slot&gt;` to quickly jump to the assigned node
<a href="https://bugs.webkit.org/show_bug.cgi?id=290679">https://bugs.webkit.org/show_bug.cgi?id=290679</a>

Reviewed by BJ Burg.

It&apos;s not always obvious where the assigned node is for a `&lt;slot&gt;` (assuming it&apos;s even nearby in the DOM).
Having a way to quickly see (and jump to) any assigned DOM nodes will help developers more easily understand the content they&apos;re debugging.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::isSlotElementWithAssignedNodes): Added.
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol):
(WebCore::InspectorCSSAgent::didChangeAssignedNodes): Added.
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeAssignedNodes): Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didChangeAssignedNodesImpl): Added.
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::enqueueSlotChangeEvent):
Introduce a new `LayoutFlag.SlotFilled` in order to (re)use the existing DOM node badging system.
This has the added benefit of debouncing changes, which is ideal since the frontend doesn&apos;t actually need the list of assigned nodes until the developer clicks on the &quot;Assigned&quot; badge.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::requestAssignedNodes): Added.
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.async requestAssignedNodes): Added.
Add a way for the frontend to request the list of assigned nodes for a given `&lt;slot&gt;`

* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):
(WI.DOMTreeElement.prototype.async _handleSlotFilledBadgeClicked): Added.
(WI.DOMTreeElement.prototype.didDismissPopover):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.event-badge-popover-content, .slot-filled-badge-popover-content): Renamed from `.event-badge-popover-content`.
(.slot-filled-badge-popover-content &gt; table): Added.
(.slot-filled-badge-popover-content &gt; table &gt; tr + tr &gt; td): Added.
(.slot-filled-badge-popover-content &gt; table &gt; tr &gt; td:first-child): Added.
Show an &quot;Assigned&quot; badge for `&lt;slot&gt;` that have assigned nodes.
Clicking it will either select the (single) assigned DOM node or show a `WI.Popover` containing a list of all assigned DOM nodes (if there&apos;s more than one).

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-SlotFilled.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-SlotFilled-expected.txt: Added.
* LayoutTests/inspector/dom/requestAssignedNodes.html: Added.
* LayoutTests/inspector/dom/requestAssignedNodes-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/292996@main">https://commits.webkit.org/292996@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a06a3bac7f7d0af8279db8a25d339e8e6c29c5ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7495 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/102759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99698 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25731 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/102759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31575 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100656 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/102759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/13083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6190 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47624 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/90329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104761 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96275 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24733 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25105 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82863 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20865 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27422 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18328 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24694 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119901 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24516 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/27830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26090 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -1044,6 +1045,19 @@ static bool hasJSEventListener(Node& node)
     return false;
 }
 
+static bool hasAssignedNodes(Node& node)
```
Since "assigned nodes" is an obscure term...

Naming suggestion: `isSlotElementWithAssignedNodes`

burg commented on Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
```diff
@@ -648,6 +649,32 @@ Inspector::Protocol::ErrorStringOr<void> InspectorDOMAgent::requestChildNodes(In
     return { };
 }
 
+Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::DOM::NodeId>>> InspectorDOMAgent::requestAssignedNodes(Inspector::Protocol::DOM::NodeId nodeId)
```
You can use the newer `Inspector::CommandResult<>` in place of `ErrorStringOr<>`

burg commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2207,6 +2215,54 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+
+        let tableContainer = contentElement.appendChild(document.createElement("table"));
+
+        let assignedNodes = await this.representedObject.requestAssignedNodes();
```
Hmm.. should we await fetching the nodes first before creating the Popover?

burg commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2255,8 +2311,17 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
     {
-        if (popover === this._eventBadgePopover)
+        switch (popover) {
+        case this._eventBadgePopover:
```
This reads a little weird to me, but I guess its ok?

dcrousso committed:
Web Inspector: Elements: show a badge for `<slot>` to quickly jump to the assigned node
https://bugs.webkit.org/show_bug.cgi?id=290679

Reviewed by BJ Burg.

It's not always obvious where the assigned node is for a `<slot>` (assuming it's even nearby in the DOM).
Having a way to quickly see (and jump to) any assigned DOM nodes will help developers more easily understand the content they're debugging.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::isSlotElementWithAssignedNodes): Added.
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol):
(WebCore::InspectorCSSAgent::didChangeAssignedNodes): Added.
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeAssignedNodes): Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didChangeAssignedNodesImpl): Added.
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::HTMLSlotElement::enqueueSlotChangeEvent):
Introduce a new `LayoutFlag.SlotFilled` in order to (re)use the existing DOM node badging system.
This has the added benefit of debouncing changes, which is ideal since the frontend doesn't actually need the list of assigned nodes until the developer clicks on the "Assigned" badge.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::requestAssignedNodes): Added.
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.async requestAssignedNodes): Added.
Add a way for the frontend to request the list of assigned nodes for a given `<slot>`

* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):
(WI.DOMTreeElement.prototype.async _handleSlotFilledBadgeClicked): Added.
(WI.DOMTreeElement.prototype.didDismissPopover):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.event-badge-popover-content, .slot-filled-badge-popover-content): Renamed from `.event-badge-popover-content`.
(.slot-filled-badge-popover-content > table): Added.
(.slot-filled-badge-popover-content > table > tr + tr > td): Added.
(.slot-filled-badge-popover-content > table > tr > td:first-child): Added.
Show an "Assigned" badge for `<slot>` that have assigned nodes.
Clicking it will either select the (single) assigned DOM node or show a `WI.Popover` containing a list of all assigned DOM nodes (if there's more than one).

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-SlotFilled.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-SlotFilled-expected.txt: Added.
* LayoutTests/inspector/dom/requestAssignedNodes.html: Added.
* LayoutTests/inspector/dom/requestAssignedNodes-expected.txt: Added.

Canonical link: https://commits.webkit.org/292996@main
